ext/curl/share.c

```diff
@@ -134,6 +136,159 @@ PHP_FUNCTION(curl_share_strerror)
+
+		if (failed) {
+			zend_argument_type_error(1, "must contain only integer values, %s given", zend_zval_value_name(entry));
+			goto error;
+		}
+
+		switch (option) {
+			// Disallowed options
+			case CURL_LOCK_DATA_COOKIE:
+				zend_argument_value_error(1, "must not contain CURL_LOCK_DATA_COOKIE because it is unsafe");
+				goto error;
+
+			// Allowed options
+			case CURL_LOCK_DATA_DNS:
+				persistent_id |= 1 << 0;
+				break;
+			case CURL_LOCK_DATA_SSL_SESSION:
+				persistent_id |= 1 << 1;
+				break;
+			case CURL_LOCK_DATA_CONNECT:
+				persistent_id |= 1 << 2;
+				break;
+			case CURL_LOCK_DATA_PSL:
+				persistent_id |= 1 << 3;
+				break;
+
+			// Unknown options
+			default:
+				zend_argument_value_error(1, "must contain only CURL_LOCK_DATA_* constants");
+				goto error;
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	// Construct a property field for the CurlSharePersistentHandle object with the enabled share options.
+	zval share_opts_prop;
+	array_init(&share_opts_prop);
+
+	if (persistent_id & 1 << 0) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_DNS);
+	}
+
+	if (persistent_id & 1 << 1) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_SSL_SESSION);
+	}
+
+	if (persistent_id & 1 << 2) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_CONNECT);
+	}
+
+	if (persistent_id & 1 << 3) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_PSL);
+	}
+
+	object_init_ex(return_value, curl_share_persistent_ce);
+
+	zend_update_property(
+	    curl_share_persistent_ce,
+		Z_OBJ_P(return_value),
+		"options",
+		sizeof("options") - 1,
+		&share_opts_prop
+	);
+
+	zval_ptr_dtor(&share_opts_prop);
+
+	sh = Z_CURL_SHARE_P(return_value);
+
+	if (persistent_id) {
+		zval *persisted = zend_hash_index_find(&CURL_G(persistent_curlsh), persistent_id);
+
+		if (persisted) {
+			sh->share = Z_PTR_P(persisted);
+
+			return;
+		}
+	}
+
+	// We could not find an existing share handle, so we'll have to create one.
+	sh->share = curl_share_init();
+
+	// Apply $share_options to the handle.
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(share_opts), entry) {
+		ZVAL_DEREF(entry);
+
+		error = curl_share_setopt(sh->share, CURLSHOPT_SHARE, zval_get_long(entry));
+
+		if (error != CURLSHE_OK) {
+			zend_throw_exception_ex(
+				NULL,
+				0,
+				"Could not construct persistent cURL share handle: %s",
+				curl_share_strerror(error)
+			);
+
+			goto error;
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	zend_hash_index_add_new_ptr(
+		&CURL_G(persistent_curlsh),
+		persistent_id,
+		sh->share
+	);
+
+	return;
+
+ error:
+	if (sh && sh->share) {
+		curl_share_cleanup(sh->share);
+	}
+
+	RETURN_THROWS();
+}
+
+/**
+ * Free a persistent curl share handle from the module global HashTable.
+ *
+ * See PHP_GINIT_FUNCTION in ext/curl/interface.c.
+ */
+void curl_share_free_persistent_curlsh(zval *data)
+{
+	CURLSH *handle = Z_PTR_P(data);
+
+	if (!handle) {
+		return;
+	}
+
+	curl_share_cleanup(handle);
```

Comment:
```suggestion
	CURLSH *handle = Z_PTR_P(data);

	curl_share_cleanup(handle);
```
> Passing in a NULL pointer in share_handle makes this function return immediately with no action.

(https://curl.se/libcurl/c/curl_share_cleanup.html)

Reply:
Nice catch!